### ISSUE TITLE
`StyletronProvider` issue with TypeScript 2.4.1

### ISSUE THREAD

**Issue by mattbasta** [rtsao/styletron#153]
After upgrading to TypeScript 2.4.1 (from 2.3), I started getting this error:

```
ERROR in [at-loader] ./src/framing.tsx:43:9 
    TS2605: JSX element type 'StyletronProvider' is not a constructor function for JSX elements.
  Types of property 'setState' are incompatible.
    Type '{ <K extends never>(f: (prevState: void, props: { styletron: StyletronServer | StyletronClient; }...' is not assignable to type '{ <K extends never>(f: (prevState: {}, props: any) => Pick<{}, K>, callback?: (() => any) | undef...'.
      Types of parameters 'f' and 'f' are incompatible.
        Types of parameters 'prevState' and 'prevState' are incompatible.
          Type 'void' is not assignable to type '{}'.
```

The code that it comes from is fairly mundane:

```typescript
import {StyletronProvider} from 'styletron-react';
// ...
const styletron = new Styletron();
const root = (
    <StyletronProvider styletron={styletron}>
        {elem}
    </StyletronProvider>
);
```

Based on my understanding of TS 2.4, I think [this](https://github.com/rtsao/styletron/blob/master/packages/styletron-react/index.d.ts#L9) should be updated like this:

```diff
declare namespace StyletronReact {
-  export class StyletronProvider extends React.Component<{styletron: StyletronServer | StyletronClient}, void> {}
+  export class StyletronProvider extends React.Component<{styletron: StyletronServer | StyletronClient}, {}> {}

  export function styled<TProps>(
    base: React.StatelessComponent<TProps & ClassNameProp>,
    style: Style<TProps>): Component<TProps>;
```

Based on [React's type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L209), it would seem that state cannot be `void`.

I haven't tested whether this change would negatively impact TypeScript 2.3 or earlier.

**Comment by ehellman** [comment]
I'm getting the same issue with 2.5.8 and TypeScript 2.4.1. Tried downgrading to TypeScript 2.3 and got the same issue.

**Comment by faddee** [comment]
It's due to some changes in latest `@types/react`. It's actually patched in master for the next major release. But we should definitely patch this for v2 too. This version is working for me though: `@types/react@15.0.25`

**Comment by ehellman** [comment]
@faddee Those types work for you on which version of TypeScript and styletron?

**Comment by faddee** [comment]
`typescript@2.3.4` and `styletron-react@2.5.8`. There's a fix coming soon for v2.

**Comment by rtsao** [comment]
Released in `styletron-react@2.5.10` Thanks @faddee!
